src/Front/Ast.hs

```diff
@@ -108,88 +110,69 @@ data Literal
+  pretty (DefPat b body) = pretty b <+> pretty "=" <+> pretty body
 
 instance Pretty Bind where
-  pretty = undefined
-  {-
-  pretty (Bind id mty) =
-    let prettyId = pretty id
-    in  case mty of
-          Just ty -> prettyId <+> pretty ':' <+> pretty ty
-          Nothing -> prettyId
-  pretty (TupBind binds mty) =
-    let prettyTup = hsep (punctuate comma $ map pretty binds)
-    in  case mty of
-          Just ty -> parens prettyTup <+> pretty ':' <+> pretty ty
-          Nothing -> prettyTup
-  -}
-
-  {-
-  pretty (PId      s) = pretty s
-  pretty (PLiteral l) = pretty l
-  pretty PWildcard    = pretty '_'
-  pretty (PAs  v p )  = pretty v <> pretty '@' <> pretty p
-  pretty (PCon c ps)  = pretty c <+> hsep (map pretty ps)
-  -}
+  pretty (Bind pat []) = pretty pat -- TODO(hans) why is ty a list?
+  pretty (Bind pat [ty]) = pretty pat <+> pretty ":" <+> pretty ty
+  pretty (Bind pat tys) =
+    pretty pat
+      <+> pretty ": ["
+      <+> hsep (punctuate comma $ map pretty tys)
+      <+> pretty "]"
+
+instance Pretty BindPat where
+  pretty BindWildcard   = pretty '_'
+  pretty (BindId  s   ) = pretty s
+  pretty (BindLit l   ) = pretty l
+  pretty (BindAs b p  ) = pretty b <> pretty '@' <> pretty p -- TODO(hans): Parens around p?
+  pretty (BindTup bs  ) = parens $ hsep (punctuate comma $ map pretty bs)
+  pretty (BindCon c ps) = pretty c <+> hsep (map pretty ps)
 
 instance Pretty TypFn where
-  pretty = undefined
+  pretty (TypReturn t) = pretty "->" <+> pretty t
+  pretty (TypProper t) = pretty ":" <+> pretty t
+  pretty TypNone       = pretty ": ()" -- TODO(hans): What is this?
 
 instance Pretty Typ where
-  pretty = undefined
-    {-
   pretty (TCon id                   ) = pretty id
   pretty (TApp (TCon "[]") t2) = pretty "[" <> parens (pretty t2) <> pretty "]"
   pretty (TApp (TCon "&" ) t2       ) = pretty "&" <> parens (pretty t2)
   pretty (TApp t           (TCon id)) = pretty t <+> pretty id
   pretty (TApp t1          t2       ) = pretty t1 <+> parens (pretty t2)
-  pretty (TTuple tys) =
-    pretty "(" <> hsep (punctuate comma $ map pretty tys) <> pretty ")"
-  pretty (TArrow t1 t2) = pretty t1 <+> pretty "->" <+> pretty t2
-  -}
+  pretty (TTuple tys) = parens $ hsep (punctuate comma $ map pretty tys)
+  pretty (TArrow t1 t2              ) = pretty t1 <+> pretty "->" <+> pretty t2
 
 
 instance Pretty Expr where
-  pretty = undefined
-    {-
-  pretty (Id      id         ) = pretty id
-  pretty (Literal l          ) = pretty l
+  pretty (Id  id             ) = pretty id
+  pretty (Lit l              ) = pretty l
   pretty (Apply    (Id id) e ) = pretty id <+> pretty e
   pretty (Apply    e1      e2) = parens (pretty e1) <+> pretty e2
   pretty (OpRegion e1      r ) = parens (pretty e1 <> p r)
    where
     p EOR             = emptyDoc
     p (NextOp s e r') = space <> pretty s <+> pretty e <> p r'
-  pretty (As v e) = pretty v <> pretty '@' <> pretty e
-  pretty NoExpr   = emptyDoc
+  pretty NoExpr = emptyDoc
```

Comment:
This should either throw an error, or chuck out some very visible `pretty "NOEXPR"` token here. `pretty` shouldn't be called on `NoExpr` on its own, because it really shouldn't appear anywhere aside from in the `else` branch of the `IfElse` construct.

Reply:
My concern with just returning an empty document here is that, from a malformed AST, you will produce a malformed document that doesn't parse because of a quietly missing token, and it will be annoying to debug. So if you come across such a malformed AST, it should either fail early or fail loudly.